File: src/main/features/uploader/UploaderService.ts
import { clipboard } from 'electron'
import { v4 as uuidv4 } from 'uuid'
import { configManager } from '../../core/ConfigManager'
import { logger } from '../../core/Logger'
import { compressImage, getImageMeta } from './ImageCompressor'
import { githubUploader } from './GitHubUploader'
import { uploaderDataManager } from './UploaderDataManager'
import type { UploaderConfig, UploadRecord, UploadResult } from '@shared/types'
import type { ImageMetaMain, CompressResultMain } from './types'

class UploaderService {
  private config: UploaderConfig | null = null
  private pendingImage: { buffer: number[]; filename: string } | null = null

  init(): void {
    uploaderDataManager.init()
    this.loadConfig()
    configManager.on('config:updated', () => this.loadConfig())
  }

  private loadConfig(): void {
    const cfg = configManager.get() as { uploader?: UploaderConfig }
    if (cfg.uploader) {
      this.config = cfg.uploader
      if (this.config.github.token) {
        githubUploader.configure(this.config.github)
      }
    }
  }

  getConfig(): UploaderConfig | null {
    return this.config
  }

  async getClipboardImage(): Promise<ImageMetaMain | null> {
    try {
      const image = clipboard.readImage()
      if (image.isEmpty()) {
        return null
      }
      const buffer = image.toPNG()
      const meta = await getImageMeta(buffer)
      return {
        buffer,
        format: meta.format,
        width: meta.width,
        height: meta.height,
        size: buffer.length
      }
    } catch (err) {
      logger.error('Failed to read clipboard image', err)
      return null
    }
  }

  async compressImage(
    buffer: Buffer,
    quality: number,
    format: 'auto' | 'webp' | 'jpeg' | 'png' = 'auto'
  ): Promise<CompressResultMain & { outputFormat: string }> {
    return compressImage(buffer, quality, format)
  }

  async upload(
    buffer: Buffer,
    filename: string,
    path: string,
    meta: { width: number; height: number; format: string; originalSize: number }
  ): Promise<UploadResult & { record?: UploadRecord }> {
    if (!this.config) {
      return { success: false, error: 'Uploader not configured' }
    }
    const result = await githubUploader.upload(buffer, path, filename, this.config.cdn.baseUrl)
    if (!result.success) {
      return result
    }

    const record: UploadRecord = {
      id: uuidv4(),
      filename,
      originalName: filename,
      timestamp: new Date().toISOString(),
      originalSize: meta.originalSize,
      compressedSize: buffer.length,
      width: meta.width,
      height: meta.height,
      format: meta.format as 'png' | 'jpeg' | 'webp' | 'gif',
      path,
      cdnUrl: result.cdnUrl!,
      sha: result.sha!
    }
    // Save with thumbnail if enabled
    if (this.config.cacheThumbnails) {
      await uploaderDataManager.addRecord(record, buffer)
    } else {
      await uploaderDataManager.addRecord(record)
    }
    return { ...result, record }
  }

  async deleteRecord(id: string): Promise<{ success: boolean; error?: string }> {
    const record = uploaderDataManager.getRecord(id)
    if (!record) {
      return { success: false, error: 'Record not found' }
    }
    const fullPath = record.path ? `${record.path}/${record.filename}` : record.filename
    const result = await githubUploader.delete(fullPath, record.sha)
    if (result.success) {
      uploaderDataManager.removeRecord(id)
    }
    return result
  }

  getHistory(): UploadRecord[] {
    return uploaderDataManager.getHistory()
  }

  getRecentPaths(): string[] {
    return uploaderDataManager.getRecentPaths()
  }

  copyToClipboard(text: string): void {
    clipboard.writeText(text)
  }

  setPendingImage(data: { buffer: number[]; filename: string }): void {
    this.pendingImage = data
  }

  getPendingImage(): { buffer: number[]; filename: string } | null {
    const image = this.pendingImage
    this.pendingImage = null // Clear after retrieval
    return image
  }

  getThumbnail(id: string): Buffer | null {
    const thumbPath = uploaderDataManager.getThumbnailPath(id)
    if (!thumbPath) return null
    try {
      return require('fs').readFileSync(thumbPath)
    } catch {
      return null
    }
  }
}

export const uploaderService = new UploaderService()

File: nexus-swift/Nexus/Features/Uploader/UploaderService.swift
import Foundation
import AppKit
import Combine

/// Coordinates the uploader's pieces: live config → GitHubUploader,
/// compression on demand, clipboard helpers, and a one-shot "pending image"
/// slot the AppDelegate uses when the user drags onto the menu bar icon.
@MainActor
final class UploaderService: ObservableObject {
    /// Read-only signal: nil means "no GitHub credentials configured".
    @Published private(set) var isConfigured: Bool = false
    /// Set by the menu bar drag receiver; consumed by the next mount of
    /// UploaderPopoverView. Reset to nil after consumption.
    @Published var pendingImage: PendingImage?

    let repository: UploaderRepository
    private let config: ConfigService
    private let github = GitHubUploader()

    private var configCancellable: AnyCancellable?

    init(repository: UploaderRepository, config: ConfigService) {
        self.repository = repository
        self.config = config
    }

    func bootstrap() async {
        await repository.bootstrap()
        configCancellable = config.$config
            .map(\.uploader)
            .removeDuplicates()
            .sink { [weak self] cfg in
                Task { @MainActor in await self?.applyConfig(cfg) }
            }
        await applyConfig(config.config.uploader)
    }

    func shutdown() async {
        // Nothing async to drain — DB writes are commit-durable; thumbnails
        // are written outside transactions but each is a single file op.
    }

    // MARK: - Public

    nonisolated func compress(_ data: Data, quality: Int, format: OutputFormat) throws -> CompressResult {
        try ImageCompressor.compress(data, quality: quality, format: format)
    }

    nonisolated func meta(_ data: Data) throws -> ImageMeta {
        try ImageCompressor.meta(data)
    }

    /// Compress (using the active config) + PUT to GitHub + record to DB +
    /// copy CDN URL to clipboard. Returns the persisted record on success.
    func upload(originalData: Data,
                compressedData: Data,
                meta: ImageMeta,
                outputFormat: ImageFormat,
                filename: String,
                path: String) async throws -> UploadRecord {
        let cfg = config.config.uploader
        guard !cfg.github.token.isEmpty,
              !cfg.github.owner.isEmpty,
              !cfg.github.repo.isEmpty
        else { throw UploaderError.notConfigured }

        let outcome = try await github.upload(
            data: compressedData,
            path: path,
            filename: filename,
            cdnBaseUrl: cfg.cdn.baseUrl
        )

        let record = UploadRecord(
            id: UUID().uuidString,
            filename: filename,
            originalName: filename,
            timestamp: Date(),
            originalSize: originalData.count,
            compressedSize: compressedData.count,
            width: meta.width,
            height: meta.height,
            format: outputFormat,
            path: path.isEmpty ? nil : path,
            cdnUrl: outcome.cdnUrl,
            sha: outcome.sha,
            githubOwner: cfg.github.owner,
            githubRepo: cfg.github.repo,
            githubBranch: cfg.github.branch
        )
        // Cache thumbnail when configured. Use the *compressed* bytes — we
        // already paid to encode them, and the thumbnail of a near-final
        // image is closer to what users see in the URL.
        var thumbnail: Data?
        if cfg.cacheThumbnails {
            thumbnail = try? ImageCompressor.thumbnail(compressedData,
                                                       size: UploaderRepository.thumbnailSize)
        }
        await repository.add(record, thumbnailData: thumbnail)
        copyToClipboard(outcome.cdnUrl)
        return record
    }

    // MARK: - Clipboard

    /// PNG bytes from the system pasteboard if the topmost item is an image.
    func clipboardImage() -> Data? {
        let pb = NSPasteboard.general
        if let png = pb.data(forType: .png) { return png }
        if let tiff = pb.data(forType: .tiff),
           let rep = NSBitmapImageRep(data: tiff),
           let png = rep.representation(using: .png, properties: [:]) {
            return png
        }
        return nil
    }

    func copyToClipboard(_ text: String) {
        let pb = NSPasteboard.general
        pb.clearContents()
        pb.setString(text, forType: .string)
    }

    // MARK: - Pending image hand-off (used by status bar drag receiver)

    func setPending(_ pending: PendingImage) {
        pendingImage = pending
    }

    /// Consume the pending image (read-once); used by UploaderPopoverView on appear.
    func takePending() -> PendingImage? {
        let p = pendingImage
        pendingImage = nil
        return p
    }

    // MARK: - Private

    private func applyConfig(_ cfg: UploaderConfig) async {
        await github.configure(cfg.github)
        let configured = !cfg.github.token.isEmpty
            && !cfg.github.owner.isEmpty
            && !cfg.github.repo.isEmpty
        if configured != isConfigured {
            isConfigured = configured
            Log.uploader.info(configured ? "configured for \(cfg.github.owner)/\(cfg.github.repo)@\(cfg.github.branch)" : "not configured")
        }
    }
}
